graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -83,33 +73,49 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
         this.streamProcessingTimeout = streamFaultManager.getStreamProcessingTimeout();
         this.fingerprint = new StreamListFingerprint(streams).getFingerprint();
 
-        for (final Stream stream : streams) {
-            for (final StreamRule streamRule : stream.getStreamRules()) {
+        this.rulesList = Lists.newArrayList();
```

Comment:
This list should be initialized with the (maximum) size `presenceRules.size()  + exactRules .size() + greaterRules.size() + smallerRules.size() + regexRules.size()`.